File: volo-api/internal/intent/parser.go
package intent

import (
	"regexp"
	"strings"
)

// Result represents a parsed intent.
type Result struct {
	Action     string
	Target     *string
	Query      *string
	Confidence float64
	ExecuteURL string
}

// Parser is the rule-based intent parser (server-side mirror of the extension parser).
type Parser struct {
	siteMap   map[string]string
	searchMap map[string]string // site → search URL template (%s = query)
	aliases   map[string]string
}

func NewParser() *Parser {
	return &Parser{
		siteMap: map[string]string{
			"youtube":       "https://www.youtube.com",
			"google":        "https://www.google.com",
			"github":        "https://www.github.com",
			"twitter":       "https://www.x.com",
			"x":             "https://www.x.com",
			"reddit":        "https://www.reddit.com",
			"spotify":       "https://open.spotify.com",
			"discord":       "https://discord.com",
			"gmail":         "https://mail.google.com",
			"netflix":       "https://www.netflix.com",
			"twitch":        "https://www.twitch.tv",
			"linkedin":      "https://www.linkedin.com",
			"stackoverflow": "https://stackoverflow.com",
		},
		searchMap: map[string]string{
			"youtube": "https://www.youtube.com/results?search_query=%s",
			"google":  "https://www.google.com/search?q=%s",
			"github":  "https://github.com/search?q=%s",
			"spotify": "https://open.spotify.com/search/%s",
			"reddit":  "https://www.reddit.com/search/?q=%s",
		},
		aliases: map[string]string{
			"tube": "youtube",
			"yt":   "youtube",
			"gh":   "github",
			"tw":   "twitter",
			"so":   "stackoverflow",
		},
	}
}

var (
	playPattern       = regexp.MustCompile(`^open\s+(\w+)\s+play\s+(.+)$`)
	openSearchPattern = regexp.MustCompile(`^open\s+(\w+)\s+(.+)$`)
	openPattern       = regexp.MustCompile(`^(?:open|go to|goto)\s+(.+)$`)
	searchPattern     = regexp.MustCompile(`^(?:search for|search|look up|find)\s+(.+)$`)
)

// Parse takes a transcript and returns the parsed intent.
func (p *Parser) Parse(raw string) Result {
	text := strings.ToLower(strings.TrimSpace(raw))

	// Browser control
	if r := p.parseBrowserControl(text); r != nil {
		return *r
	}

	// "open [site] play [query]"
	if m := playPattern.FindStringSubmatch(text); m != nil {
		site := p.findSite(m[1])
		if site != "" {
			query := m[2]
			return Result{
				Action:     "open-and-play",
				Target:     &site,
				Query:      &query,
				Confidence: 0.92,
				ExecuteURL: p.buildSearchURL(site, query),
			}
		}
	}

	// "open [site] [query]"
	if m := openSearchPattern.FindStringSubmatch(text); m != nil {
		site := p.findSite(m[1])
		if site != "" {
			query := m[2]
			return Result{
				Action:     "open-and-search",
				Target:     &site,
				Query:      &query,
				Confidence: 0.9,
				ExecuteURL: p.buildSearchURL(site, query),
			}
		}
	}

	// "open [site]" or "go to [site]"
	if m := openPattern.FindStringSubmatch(text); m != nil {
		input := strings.TrimSpace(m[1])
		// Check for URL (has dot)
		if strings.Contains(input, ".") {
			url := input
			if !strings.HasPrefix(url, "http") {
				url = "https://" + url
			}
			return Result{
				Action:     "navigate",
				Target:     &url,
				Confidence: 0.85,
				ExecuteURL: url,
			}
		}
		site := p.findSite(input)
		if site != "" {
			url := p.siteMap[site]
			return Result{
				Action:     "navigate",
				Target:     &site,
				Confidence: 0.92,
				ExecuteURL: url,
			}
		}
	}

	// "search [query]"
	if m := searchPattern.FindStringSubmatch(text); m != nil {
		query := m[1]
		url := strings.ReplaceAll(p.searchMap["google"], "%s", strings.ReplaceAll(query, " ", "+"))
		return Result{
			Action:     "search",
			Query:      &query,
			Confidence: 0.9,
			ExecuteURL: url,
		}
	}

	// Fallback: treat as search
	query := text
	url := strings.ReplaceAll(p.searchMap["google"], "%s", strings.ReplaceAll(query, " ", "+"))
	return Result{
		Action:     "search",
		Query:      &query,
		Confidence: 0.6,
		ExecuteURL: url,
	}
}

func (p *Parser) parseBrowserControl(text string) *Result {
	controls := map[string]string{
		"go back":        "back",
		"back":           "back",
		"go forward":     "forward",
		"forward":        "forward",
		"close tab":      "close-tab",
		"close this tab": "close-tab",
		"new tab":        "new-tab",
		"open new tab":   "new-tab",
	}

	for pattern, action := range controls {
		if text == pattern {
			return &Result{
				Action:     "browser-control",
				Query:      &action,
				Confidence: 0.95,
			}
		}
	}

	if strings.HasPrefix(text, "scroll down") {
		q := "scroll-down"
		return &Result{Action: "browser-control", Query: &q, Confidence: 0.9}
	}
	if strings.HasPrefix(text, "scroll up") {
		q := "scroll-up"
		return &Result{Action: "browser-control", Query: &q, Confidence: 0.9}
	}

	return nil
}

func (p *Parser) findSite(input string) string {
	clean := strings.ToLower(regexp.MustCompile(`[^a-z0-9]`).ReplaceAllString(input, ""))

	// Direct match
	if _, ok := p.siteMap[clean]; ok {
		return clean
	}

	// Alias match
	if alias, ok := p.aliases[clean]; ok {
		return alias
	}

	// Partial match
	for key := range p.siteMap {
		if strings.Contains(key, clean) || strings.Contains(clean, key) {
			return key
		}
	}

	return ""
}

func (p *Parser) buildSearchURL(site, query string) string {
	template, ok := p.searchMap[site]
	if !ok {
		// Fall back to site base URL
		if url, ok := p.siteMap[site]; ok {
			return url
		}
		return ""
	}
	return strings.ReplaceAll(template, "%s", strings.ReplaceAll(query, " ", "+"))
}

File: volo-extension/src/shared/parser.ts
import { CommandResult, SITE_MAP, SEARCH_MAP } from "./types";

const WAKE_WORD = "hey volo";

/**
 * Check if transcript contains the wake word.
 * Returns the command portion after the wake word, or null.
 */
export function extractAfterWakeWord(transcript: string): string | null {
  const lower = transcript.toLowerCase().trim();
  const idx = lower.indexOf(WAKE_WORD);
  if (idx === -1) return null;
  const after = lower.slice(idx + WAKE_WORD.length).trim();
  return after.length > 0 ? after : "";
}

/**
 * Check if a transcript starts with or contains the wake word.
 */
export function containsWakeWord(transcript: string): boolean {
  return transcript.toLowerCase().trim().includes(WAKE_WORD);
}

/**
 * Rule-based intent parser.
 * Handles the fast path for obvious commands.
 */
export function parseCommand(raw: string): CommandResult {
  const text = raw.toLowerCase().trim();

  // Browser control commands
  if (text === "go back" || text === "back") {
    return { action: "browser-control", query: "back", confidence: 0.95 };
  }
  if (text === "go forward" || text === "forward") {
    return { action: "browser-control", query: "forward", confidence: 0.95 };
  }
  if (text === "close tab" || text === "close this tab") {
    return { action: "browser-control", query: "close-tab", confidence: 0.95 };
  }
  if (text === "new tab" || text === "open new tab") {
    return { action: "browser-control", query: "new-tab", confidence: 0.95 };
  }
  if (text.startsWith("scroll down")) {
    return { action: "browser-control", query: "scroll-down", confidence: 0.9 };
  }
  if (text.startsWith("scroll up")) {
    return { action: "browser-control", query: "scroll-up", confidence: 0.9 };
  }

  // "open [site] play [query]" → open-and-play
  const playMatch = text.match(/^open\s+(\w+)\s+play\s+(.+)$/);
  if (playMatch) {
    const [, site, query] = playMatch;
    const target = findSite(site);
    if (target) {
      return { action: "open-and-play", target, query, confidence: 0.92 };
    }
  }

  // "open [site] [query]" → open-and-search
  const openSearchMatch = text.match(/^open\s+(\w+)\s+(.+)$/);
  if (openSearchMatch) {
    const [, site, query] = openSearchMatch;
    const target = findSite(site);
    if (target) {
      return { action: "open-and-search", target, query, confidence: 0.9 };
    }
  }

  // "open [site]" → navigate
  const openMatch = text.match(/^(?:open|go to|goto)\s+(.+)$/);
  if (openMatch) {
    const input = openMatch[1].trim();
    // Check if it looks like a URL (has a dot) before trying site matching
    if (input.includes(".")) {
      const url = input.startsWith("http") ? input : `https://${input}`;
      return { action: "navigate", target: url, confidence: 0.85 };
    }
    const site = findSite(input);
    if (site) {
      return { action: "navigate", target: site, confidence: 0.92 };
    }
  }

  // "search [query]" or "search for [query]"
  const searchMatch = text.match(/^(?:search for|search|look up|find)\s+(.+)$/);
  if (searchMatch) {
    return { action: "search", query: searchMatch[1], confidence: 0.9 };
  }

  // Fallback: treat entire text as a search query
  return { action: "search", query: text, confidence: 0.6 };
}

/**
 * Find a site key from user input (fuzzy).
 */
function findSite(input: string): string | undefined {
  const clean = input.toLowerCase().replace(/[^a-z0-9]/g, "");

  // Direct match
  if (SITE_MAP[clean]) return clean;

  // Fuzzy: "tube" → youtube, "yt" → youtube
  const aliases: Record<string, string> = {
    tube: "youtube",
    yt: "youtube",
    gh: "github",
    tw: "twitter",
    ig: "instagram",
    so: "stackoverflow",
  };
  if (aliases[clean]) return aliases[clean];

  // Partial match
  for (const key of Object.keys(SITE_MAP)) {
    if (key.includes(clean) || clean.includes(key)) return key;
  }

  return undefined;
}

/**
 * Build the execution URL for a command.
 */
export function buildExecutionUrl(command: CommandResult): string | null {
  switch (command.action) {
    case "search":
      return `https://www.google.com/search?q=${encodeURIComponent(command.query || "")}`;

    case "navigate":
      if (command.target && SITE_MAP[command.target]) {
        return SITE_MAP[command.target];
      }
      return command.target || null;

    case "open-and-search":
    case "open-and-play":
      if (command.target && command.query && SEARCH_MAP[command.target]) {
        return SEARCH_MAP[command.target](command.query);
      }
      if (command.target && SITE_MAP[command.target]) {
        return SITE_MAP[command.target];
      }
      return null;

    default:
      return null;
  }
}
